### PR TITLE
[#927] Add user guide for limiting connections per tenant.

### PR DESCRIPTION
**Pull request by kaniyan**

### CODE REVIEW

File: site/content/concepts/tenant-limits.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+## Connections Limit
+
+At the moment Hono supports only the limit on the device connections. Before accepting a new connection request from a device, the corresponding Protocol adapter checks if the number of existing connections for the given Tenant exceeds the configured limit or not. If exceeds, then the connection request is declined. 
+
+The MQTT and AMQP protocol adapters keep the connections longer opened than their counterparts such as HTTP. Thereby the MQTT and AMQP adapters are enabled to check the connection limits before accepting any new connection to a device.
+
+## Tenant Configuration
+The limits for a tenant can be set using the tenant configuration. Please refer to the [Tenant API]({{< relref "/api/Tenant-API.md#request-payload" >}}) for more details.
+
+
```
Comment (sophokles73) [inline site/content/concepts/tenant-limits.md]: IMHO this whole text should be incorporated into deployment/resource-limitation.md instead

File: site/content/api/Tenant-API.md [hunk 2]
```diff
@@ -284,6 +284,7 @@ The table below provides an overview of the standard members defined for the JSO
...
 | *trusted-ca*             | *no*      | *JSON object* | `-`          | The trusted certificate authority to use for validating certificates presented by devices of the tenant for authentication purposes. See [Trusted Certificate Authority Format]({{< relref "#trusted-ca-format" >}}) for a definition of the content model of the object. |
 | *adapters*               | *no*      | *array*       | `-`          | A list of configuration options valid for certain adapters only. The format of a configuration option is described here [Adapter Configuration Format]({{< relref "#adapter-configuration-format" >}}). **NB** If the element is provided then the list MUST NOT be empty. **NB** Only a single entry per *type* is allowed. If multiple entries for the same *type* are present it is handled as an error. **NB** If the element is omitted then all adapters are *enabled* in their default configuration. |
+| *limits*                 | *no*      | *JSON object* | `-`          | The limits such as the maximum number of connections can be set. The format of a configuration option is described here [Limits Configuration Format]({{< relref "#limits-configuration-format" >}}).|
```
Comment (sophokles73) [inline site/content/api/Tenant-API.md]: IMHO just `limits` is a little too unspecific for users to understand what it is about. What about `resource-limits`? Also, the type should be just *array* (analogous to field *adapters*).

File: site/content/api/Tenant-API.md [hunk 3]
```diff
@@ -376,6 +377,12 @@ Protocol adapters SHOULD use the configuration properties set for a tenant when
 
 The JSON object MAY contain an arbitrary number of additional members with arbitrary names of either scalar or complex type.
 
+## Limits Configuration Format
+The table below provides an overview of the members defined for the *limits* JSON object:
+
+| Name                     | Mandatory | Type          | Default Value | Description |
+| :------------------------| :-------: | :------------ | :------------ | :---------- |
+| *max-connections*        | *no*      | *long*      | *-1*          | The maximum number of connections to devices to be allowed. Default value -1 indicates that no limit is set.|
```
Comment (sophokles73) [inline site/content/api/Tenant-API.md]: I don't think that there is a type *long* in JSON, is there? So this should probably be *number*.

File: site/content/user-guide/mqtt-adapter.md [hunk 4]
```diff
@@ -37,6 +37,9 @@ When running the Hono example installation as described in the [Getting Started
 There is a subtle difference between the *device identifier* (*device-id*) and the *auth-id* a device uses for authentication. See [Device Identity]({{< ref "/concepts/device-identity.md" >}}) for a discussion of the concepts.
 {{% /note %}}
 
+## Connection Limits
+After successful authentication, the Mqtt adapter verifies if the maximum number of connections exceeds the configured [limits] ({{< ref "/concepts/tenant-limits.md" >}}) or not. If exceeded then a return code `0x05` indicating `Connection Refused: not authorised` is sent back.
```
Comment (sophokles73) [inline site/content/user-guide/mqtt-adapter.md]: please use all uppercase for *MQTT*

File: site/content/admin-guide/limits-config.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+weight = 325
++++
+
+To enable the limits based on the default prometheus based implementation, the parameters in the below table should be configured.
+
+| Environment Variable<br>Command Line Option | Mandatory | Default Value | Description  |
+| :------------------------------------------ | :-------: | :------------ | :------------|
+| `hono.plan.prometheusBased.host` | no | none | The host name or ip of the prometheus backend which will be used by the default implementation to retrieve metrics data.|
+| `hono.plan.prometheusBased.port` | no | 9090 | The port of the prometheus backend which will be used by the default implementation to retrieve metrics data.|
+
```
Comment (sophokles73) [inline site/content/admin-guide/limits-config.md]: this needs to go into the admin guides of the adapters that support it.

File: site/content/api/Tenant-API.md (hunk 2 above)
Comment (kaniyan) [reply site/content/api/Tenant-API.md]: @sophokles73 `resource-limits` sounds fine to me. I will also update the `TenantConstants `accordingly. I didn't get it why it should be an _array_ analogous to field _adapters_. Each tenant configuration has only one `resource-limits` config as in below _example_ and not an _array_:

```
 {
    "tenant-id": "TENANT_TEST",
    "enabled": true,
    "resource-limits": {
      "max-connections": 25,
      "max-bytes-per-month": 26214400
    },
    "adapters": [
      {
        "type": "hono-http",
        "enabled": true,
        "device-authentication-required": true
      },
      {
        "type": "hono-mqtt",
        "enabled": false,
        "device-authentication-required": true
      }
    ]
  }
```

File: site/content/concepts/tenant-limits.md (hunk 1 above)
Comment (kaniyan) [reply site/content/concepts/tenant-limits.md]: @sophokles73 IMHO it is good to provide links in the concept page on how to configure a _tenant_ with `resource-limits` and how to _enable_ and _configure_ the default Prometheus based implementation. FMPOV when someone takes a look at the concept, then this page is self-sufficient for him/her to configure and use it. In addition to it, it is also good to add the above text to the deployment/resource-limitation.md. WDYT?

File: site/content/api/Tenant-API.md (hunk 2 above)
Comment (sophokles73) [reply site/content/api/Tenant-API.md]: my bad, should be an *object*, of course.

File: site/content/api/Tenant-API.md (hunk 3 above)
Comment (kaniyan) [reply site/content/api/Tenant-API.md]: I have changed to _number_

File: site/content/user-guide/mqtt-adapter.md (hunk 4 above)
Comment (kaniyan) [reply site/content/user-guide/mqtt-adapter.md]: Done.

File: site/content/admin-guide/limits-config.md (hunk 5 above)
Comment (kaniyan) [reply site/content/admin-guide/limits-config.md]: Done.

File: site/content/concepts/tenant-limits.md (hunk 1 above)
Comment (kaniyan) [reply site/content/concepts/tenant-limits.md]: @sophokles73 Would you mind looking at the pushed changes :wink:.

File: site/content/concepts/resource-limits.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
++++
+title = "Resource limits"
+weight = 196
++++
+
+Eclipse Hono™ users can set the resource limits such as the maximum number of device connections allowed per tenant or the allowed data volume of the messages over a period of time per tenant. 
+
+Hono specifies an API `ResourceLimitChecks` that is used by the protocol adapters for the verification of the configured resource-limits. A default implementation of this API is shipped with Hono. This default implementation uses the metrics data from the Prometheus back end to verify the resource-limits if any configured. To enable and use the default Prometheus based limits implementation, please refer to the protocol adapter admin guides (configuration parameters starting with `hono.plan.prometheusBased`). If needed, Hono users can bring their own implementation of the above API and use it. The resource-limits for a tenant can be set using the tenant configuration. Please refer to the [Tenant API]({{< relref "/api/Tenant-API.md#request-payload" >}}) for more details.
```
Comment (sophokles73) [inline site/content/concepts/resource-limits.md]: IMHO we should not explicitly mention the configuration parameter names. Referring the reader to the admin guide should be sufficient and prevents duplication of this information which we will have a hard time keeping in sync ...

File: site/content/concepts/resource-limits.md [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+
+## Connections Limit
+
+At the moment Hono supports only the limit on the device connections. Before accepting a new connection request from a device, the corresponding Protocol adapter checks if the number of existing connections for the given Tenant exceeds the configured limit or not. If exceeds, then the connection request is declined.
```
Comment (sophokles73) [inline site/content/concepts/resource-limits.md]: FMPOV we should simply omit the first sentence.

File: site/content/concepts/resource-limits.md [hunk 8]
```diff
@@ -0,0 +1,18 @@
++++
+title = "Resource limits"
+weight = 196
++++
+
+Eclipse Hono™ users can set the resource limits such as the maximum number of device connections allowed per tenant or the allowed data volume of the messages over a period of time per tenant.
```
Comment (sophokles73) [inline site/content/concepts/resource-limits.md]: we usually do not use this style of *active* writing (`Hono user can set ...`) in the documentation. Instead, we usually use a *passive* style, e.g. 
> Resource limits can be set using ...

Can you please rephrase here and in the other pages in order to maintain a consistent style?

**Comment by kaniyan** [comment]
@sophokles73 I have pushed the requested changes.

**Pull request merged**